File: Computer Graphics/mid_point_circle.cpp
#include <stdio.h>
#include <conio.h>
#include <graphics.h>
#include<time.h>

void drawcircle(int xc, int yc, int x, int y)
{
    putpixel(xc + x, yc + y, RED);
    putpixel(xc - x, yc + y, BLUE);
    putpixel(xc + x, yc - y, GREEN);
    putpixel(xc - x, yc - y, RED);
    putpixel(xc + y, yc + x, GREEN);
    putpixel(xc - y, yc + x, YELLOW);
    putpixel(xc + y, yc - x, YELLOW);
    putpixel(xc - y, yc - x, YELLOW);
}

void circlemidpoint(int xc, int yc, int r)
{
    clock_t begin=clock();
    int x = 0, y = r;
    int p = 1 - r;
    while (x < y)
    {
        drawcircle(xc, yc, x, y);
        x++;
        if (p < 0)
        {
            p = p + 2 * x + 1;
        }
        else
        {
            y--;
            p = p + 2 * (x - y) + 1;
        }
        drawcircle(xc, yc, x, y);
    }
    clock_t end = clock();
    double time_spent = (double)(end - begin) / CLOCKS_PER_SEC;
    printf("Execution time: %f seconds\n", time_spent);
}

int  main()
{
    int xc, yc, r;
    int gd = DETECT, gm;
    initgraph(&gd, &gm,(char*) "");
    printf("Enter center coordinates of the circle: ");
    scanf("%d%d", &xc, &yc);
    printf("Enter radius of the circle: ");
    scanf("%d", &r);
    circlemidpoint(xc, yc, r);
    getch();
    return 0;
}

File: Computer Graphics/mid_point_circle.py
from graphics import *
import time

def putpixel(win,x, y):  
   pt = Point(x,y)
   pt.draw(win)

def drawcircle(win,xc,yc,x,y):
    putpixel(win,xc + x, yc + y)
    putpixel(win,xc - x, yc + y)
    putpixel(win,xc + x, yc - y)
    putpixel(win,xc - x, yc - y)
    putpixel(win,xc + y, yc + x)
    putpixel(win,xc - y, yc + x)
    putpixel(win,xc + y, yc - x)
    putpixel(win,xc - y, yc - x)

def circlemidpoint(xc, yc, r):
    st = time.time()
    x=0
    y=r
    p=1-r
    win = GraphWin('Bresenham\'s Line Drawing Algorithm', 600, 480)
    while x<y:
        drawcircle(win,xc, yc, x, y)
        x+=1
        if p < 0:
            p = p + 2 * x + 1
        else:
            y-=1
            p = p + 2 * (x - y) + 1
        drawcircle(win,xc, yc, x, y)

    et= time.time()
    elapsed_time = et - st
    print('Execution time:', elapsed_time, 'seconds')
    win.getMouse()
    win.close()
  
def main():
    x1 = int(input("Enter center X: "))
    y1 = int(input("Enter center Y: "))
    r = int(input("Radius: "))
    circlemidpoint(x1, y1, r)
      
if __name__ == "__main__":
   main()
